### DISCUSSION TITLE
Solving a linear program given by a separation oracle

### DISCUSSION THREAD

**Discussion by erelsgl** [cvxpy/cvxpy#1562] [Q&A]
Some linear programs have a huge number of variables or constraints, so they cannot be fully written down. Instead, they can be described using a [separation oracle](https://en.wikipedia.org/wiki/Separation_oracle). In some cases, such problems can be solved in polynomial time using the ellipsoid method. For example, [this paper](https://ieeexplore.ieee.org/document/4568405) uses this approach for finding an approximate solution to bin packing. Is there a way to solve such problems in `cvxpy`?

**Comment by SteveDiamond** [comment]
Interesting question! cvxpy does not currently support this mode of optimization. Your best bet would be to write your own ellipsoid method or find an existing solver.

**Reply by rileyjmurray** [reply]
A small note: the ellipsoid method isn't actually practical. It would probably be preferable to solve the small LP, run the separation oracle, and then construct and solve a new LP with the constraint implied by the separation oracle. (Of course this process stops if the separation oracle says the candidate solution is feasible.) The main source of inefficiency of when using CVXPY here is that we'd be recompiling a problem from scratch every time. We also lose the benefit of warm-starting each small LP solve.

Another note: my experience with these kinds of LPs is that the mathematical description of the model is relatively simple. In these situations you can state the LP in standard form like ``A @ x <= b`` for a SciPy sparse matrix ``A``, and manually add rows to ``(A, b)`` each time you run the separation oracle. My experience is that CVXPY introduces very little overhead when problems are given in that standard form. (For example, [this code](https://github.com/rileyjmurray/sageopt/blob/8f438706d8020f26c4707d0f94aa372eba6a559b/sageopt/coniclifts/problems/solvers/cvxpy.py#L16) translates a low-level representation into a CVXPY model, and it handles millions of variables without significantly increasing runtime compared to building the model directly with MOSEK's API.)

**Comment by erelsgl** [comment]
I found [this blog post](https://orinanobworld.blogspot.com/2013/11/lp-cutting-planes-in-cplex.html) explaining how it can be done in CPLEX. It can be a good idea to add an interface to it to CVXPY, but it may be hard to do.

**Reply by rileyjmurray** [reply]
That note is interesting! The main components of implementing that approach are (1) the ability to warm-start by only providing primal variables (or some-but-not-all dual variables) and (2) the ability to efficiently update a model by adding a constraint, rather than building a new model from scratch every time. Neither of these are outside the realm of possibility with CVXPY as it exists today. However, it would take a fair amount of work to implement properly. @erelsgl if you (or anyone else) would like to step up and give that a shot, do let us know!
